content/blog/2020-05-dplyr-1-0-0-last-minute-additions/index.Rmarkdown

```diff
@@ -0,0 +1,147 @@
+## `summarise()` and grouping
+
+There\'s a common confusion about the result of `summarise()`. How do you think the result of the following code will be grouped?
+
+```{r, message = FALSE}
+homeworld_species <- starwars %>% 
+  group_by(homeworld, species) %>% 
+  summarise(n = n())
+```
+
+You might be surprised to learn that it's grouped by `homeworld`:
+
+```{r}
+head(homeworld_species, 3)
+```
+
+That's because `summarise()` always peels off the last group, based on the logic that this group now occupies a single row so there's no point grouping by it. This behaviour made perfect sense to me at the time I implemented it, but it's been a long standing source of confusion amongst dplyr users (and it doesn't make sense if your summary [returns multiple rows](https://www.tidyverse.org/blog/2020/03/dplyr-1-0-0-summarise/)).
+
+Unfortunately, it would be very difficult to change this default now because a lot of code probably relies on it. Instead, we're doing the next best thing: exposing the default behaviour more explicit and making it easier to change. In dplyr 1.0.0, the code above will display a message telling you how the result has been grouped:
+
+```{r}
+homeworld_species <- starwars %>% 
+  group_by(homeworld, species) %>% 
+  summarise(n = n())
+```
+
+The text hints at how to take control of grouping and eliminate the message: a new `.groups` argument allows you to control the grouping of the result. It currently has four possible values:
+
+-   `.groups = "drop_last"` drops the last grouping level (i.e. the default behaviour sans message).
+-   `.groups = "drop"` drops all grouping levels and returns a tibble.
+-   `.groups = "keep"` preserves the grouping of the input.
+-   `.groups = "rowwise"` turns each row into its own group.
+
+If you find the default message annoying you can suppress by setting a global option:
+
+```{r, eval = FALSE}
+options(dplyr.summarise.inform = FALSE)
+```
+
+`.groups` is very new, so we've marked it as experimental, meaning that it may change in the future. Please let us know what you think of it so that we have enough data to make a decision about its future.
+
+## Row mutation
+
+Thanks to [Kirill Müller](https://krlmlr.info/), dplyr has a new experimental family of row mutation functions inspired by SQL's `UPDATE`, `INSERT`, `UPSERT`, and `DELETE`. Like the join functions, they all work with a pair of data frames:
+
+-   `rows_update(x, y)` updates existing rows `x` in with values in `y`.
+-   `rows_patch(x, y)` works like `rows_update()` but only changes `NA` values.
+-   `rows_insert(x, y)` adds new rows to `x` from from `y`.
+-   `rows_upsert(x, y)` updates existing rows in `x` and adds new rows from `y`.
+-   `rows_delete(x, y)` deletes rows in `x` that match rows in `y`.
+
+The `row_` functions match `x` and `y` using **keys**. A key is one or more variables that uniquely identifies each row. All `row_` functions check that the keys of `x` and `y` are valid (i.e. unique) before doing anything.
```

Comment:
Do we have a principle re. where to use an asterisk and/or parentheses when referencing a family of functions (e.g., `rows_` vs. `rows_*()` vs. `rows_*`)? I think the above absence of parentheses is correct because the word "functions" comes after `rows_`, but I want to generalize a rule so we can add it to the style guide.

Reply:
For `map_*()` functions we've used the format I just used. cf https://github.com/tidyverse/style/issues/154#issuecomment-624073130

Reply:
I think `map_*()` is probably slightly clearer since we can also use it like `*_join()`